Add nREPL server support

ingydotnet opened the pull request:
$ ./bin/lingy --nrepl
Starting nrepl://127.0.0.1:49098

ingydotnet committed:
Add initial nREPL server support

$ ./bin/lingy --nrepl
Starting nrepl://127.0.0.1:49098

PEZ committed:
More work on nREPL

PEZ committed:
Use patched Bencode, encoding scalars as strings

PEZ committed:
Extract prepare and send response subs

PEZ committed:
Add message and verbose logging opts

PEZ committed:
Extract ops as separate table

PEZ committed:
Support multiple clients

PEZ committed:
Keep client counters in server instance

PEZ committed:
Create an .nrepl-port file on start

PEZ committed:
Stop and cleanup on ctrl+c

PEZ committed:
Implement `close`

ingydotnet committed:
Refactor Lingy::nREPL

PEZ committed:
Actually evaluate in the eval op

ingydotnet committed:
Makefile improvements

PEZ commented:
Note to selves: We need to figure about `stdout` and `stderr`. I think that for `eval`, `load-file`, and possibly other ops, stdout/stderr should be sent with messages with the same `id` as the request and the result value, at least until the `status => 'done'` message is sent. Not sure what to do with out-of-band output, maybe that should be sent with an `id` of `'0'` or something.

ingydotnet committed:
Improvements to logging and overall code

ingydotnet committed:
More nrepl code simplification

PEZ committed:
Honor nrepl-logging option

And also allow override for log file and to use stdout

PEZ committed:
Add nrepl command line args to usage message

PEZ committed:
Use defacto standard start message

PEZ committed:
Use non-blocking read from the socket

ingydotnet committed:
Rework pez's logging modifications

Usage:
    LINGY_NREPL_LOG_FILE=- lingy --nrepl

ingydotnet committed:
Add Bencode IO::All and Data::UUID to Meta requires

ingydotnet committed:
Support 'make nrepl log=...'

If log=- then run in foreground.

ingydotnet committed:
Make nrepl server return last result if multiple forms evaluated

PEZ committed:
Remove test multi-form code

PEZ committed:
Tie stdout to out responses while evaluating

PEZ committed:
Handle stderr

PEZ committed:
Add `ns` to value response

PEZ committed:
Set *file* when eval requests has `file`

PEZ committed:
Format according to PR feedback

PEZ committed:
Set *file* as Lingy string

PEZ committed:
Clarify why we can't print something

ingydotnet committed:
Use consistent output for printing unprintables

PEZ committed:
Don't send value response if no evaluation result

PEZ committed:
Fixup after rebasing on latest nrepl branch

ingydotnet committed:
Reformat nrepl stdio capture code

Put locals in eval to better scope them to the RT->reps call.

ingydotnet committed:
Move stdio capture code to bottom of nREPL.pm

Putting code in {} block localizes the package definition.
Just like we did for Bencode.

ingydotnet committed:
Rename a couple Lingy::nREPL attributes

ingydotnet committed:
Add nREPL info to the docs

ingydotnet committed:
Update doc/lingy.md

with suggestion from @pez

Co-authored-by: Peter Strömberg <pez@pezius.com>

ingydotnet committed:
Change LINGY_NREPL_LOG_FILE to LINGY_NREPL_LOG

ingydotnet committed:
Add Peter Strömberg to authors

ingydotnet committed:
Don't run GHA tests twice for PRs pushed here
